pkg/apis/eventing/v1/trigger_validation.go

```diff
@@ -163,3 +146,94 @@ func (t *Trigger) validateInjectionAnnotation(injectionAnnotation string) *apis.
 	}
 	return nil
 }
+
+func ValidateAttributeFilters(filter *AttributesFilter) (errs *apis.FieldError) {
+	if filter == nil {
+		return nil
+	}
+	// Validate Attributes
+	for attr := range *filter.Attributes {
+		if !validAttributeName.MatchString(attr) {
+			errs = errs.Also(apis.ErrInvalidKeyName(attr, apis.CurrentField, "Attribute name must start with a letter and can only contain lowercase alphanumeric").ViaFieldKey("attributes", attr))
+		}
+	}
+	return errs
+}
+
+func ValidateExactExpression(expr *map[string]string) (errs *apis.FieldError) {
+	if expr == nil {
+		return nil
+	}
+	if len(*expr) != 1 {
+		return apis.ErrGeneric("Exact can have only one key-value", apis.CurrentField)
+	}
+	for attr := range *expr {
+		if !validAttributeName.MatchString(attr) {
+			errs = errs.Also(apis.ErrInvalidKeyName(attr, apis.CurrentField, "Attribute name must start with a letter and can only contain lowercase alphanumeric").ViaKey(attr))
+		}
+	}
+	return errs
+}
+
+func ValidatePrefixExpression(expr *map[string]string) (errs *apis.FieldError) {
+	if expr == nil {
+		return nil
+	}
+	if len(*expr) != 1 {
+		return apis.ErrGeneric("Prefix can have only one key-value", apis.CurrentField)
+	}
+	for attr := range *expr {
+		if !validAttributeName.MatchString(attr) {
+			errs = errs.Also(apis.ErrInvalidKeyName(attr, apis.CurrentField, "Attribute name must start with a letter and can only contain lowercase alphanumeric").ViaKey(attr))
+		}
+	}
+	return errs
+}
+
+func ValidateSuffixExpression(expr *map[string]string) (errs *apis.FieldError) {
+	if expr == nil {
+		return nil
+	}
+	if len(*expr) != 1 {
+		return apis.ErrGeneric("Suffix can have only one key-value", apis.CurrentField)
+	}
+	for attr := range *expr {
+		if !validAttributeName.MatchString(attr) {
+			errs = errs.Also(apis.ErrInvalidKeyName(attr, apis.CurrentField, "Attribute name must start with a letter and can only contain lowercase alphanumeric").ViaKey(attr))
+		}
+	}
+	return errs
+}
+
+func ValidateSubscriptionAPIFilters(filters *SubscriptionsAPIFiltersList, field string) (errs *apis.FieldError) {
+	if filters == nil {
+		return nil
+	}
+	if len(*filters) < 1 {
+		return apis.ErrGeneric(fmt.Sprintf("%s must contain at least one nested filter", field), apis.CurrentField)
+	}
+
+	for i, f := range *filters {
+		f := f
+		errs = errs.Also(ValidateSubscriptionAPIFilter(&f)).ViaIndex(i)
+	}
+	return errs
+}
+
+func ValidateSubscriptionAPIFilter(filter *SubscriptionsAPIFilter) (errs *apis.FieldError) {
```

Comment:
yes, you are correct. According to the [OpenAPI  spec](https://github.com/cloudevents/spec/blob/master/subscriptions-api-openapi.yaml#L170-L185) it is indeed `oneOf` .

Reply:
In case you're not familiar with `oneOf`, the way it works is a bit counter intuitive IMO.
In case that can save you time, here is how we leverage it in a few sources at TriggerMesh:

```yaml
properties:
  broker:
    type string
  filter:
    type: object
    # ...
  filters:
    type: array
    # ...
  subscriber:
    type: object
    # ...
oneOf:
- required:
  - broker
  - filter
  - subscriber
  # filters (plural) not included -> will be rejected if specified
- required:
  - broker
  - filters
  - subscriber
  # filter (singular) not included -> will be rejected if specified
```

Reply:
I just saw Knative isn't even defining `required` right now, and relies purely on the validation webhook, so feel free to ignore my previous comment about the OpenAPI notation.